File: ScapesCore/src/main/java/org/tobi29/scapes/plugins/Plugins.java
/*
 * Copyright 2012-2015 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.tobi29.scapes.plugins;

import java8.util.stream.Collectors;
import java8.util.stream.Stream;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.tobi29.scapes.block.GameRegistry;
import org.tobi29.scapes.chunk.IDStorage;
import org.tobi29.scapes.engine.utils.Streams;
import org.tobi29.scapes.engine.utils.io.filesystem.FilePath;
import org.tobi29.scapes.engine.utils.io.filesystem.FileSystemContainer;
import org.tobi29.scapes.engine.utils.io.filesystem.FileUtil;
import org.tobi29.scapes.engine.utils.io.filesystem.classpath.ClasspathPath;
import org.tobi29.scapes.engine.utils.io.filesystem.classpath.ClasspathResource;

import java.io.IOException;
import java.net.URLClassLoader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Plugins {
    private static final Logger LOGGER = LoggerFactory.getLogger(Plugins.class);
    private final List<PluginFile> files;
    private final List<Plugin> plugins = new ArrayList<>();
    private final List<Dimension> dimensions = new ArrayList<>();
    private final GameRegistry registry;
    private final URLClassLoader classLoader;
    private WorldType worldType;
    private boolean init;

    public Plugins(List<PluginFile> files, IDStorage idStorage)
            throws IOException {
        List<FilePath> paths =
                Streams.of(files).filter(file -> file.file() != null)
                        .map(PluginFile::file).collect(Collectors.toList());
        this.files = files;
        if (paths.isEmpty()) {
            classLoader = null;
            ClassLoader classLoader = Plugins.class.getClassLoader();
            PluginFile file = new PluginFile(
                    new ClasspathResource(classLoader, "Plugin.json"));
            load(file.plugin(classLoader));
        } else {
            classLoader = new PluginClassLoader(paths);
            for (PluginFile file : files) {
                load(file.plugin(classLoader));
            }
        }
        if (worldType == null) {
            throw new IOException("No world type found");
        }
        registry = new GameRegistry(idStorage, this);
    }

    public static List<PluginFile> installed(FilePath path) throws IOException {
        List<PluginFile> files = new ArrayList<>();
        FileUtil.consumeRecursive(path, file -> {
            if (FileUtil.isRegularFile(file) && FileUtil.isNotHidden(file)) {
                files.add(new PluginFile(file));
            }
        });
        files.addAll(embedded());
        return files;
    }

    public static List<PluginFile> embedded() throws IOException {
        ClasspathResource embedded =
                new ClasspathResource(Plugins.class.getClassLoader(),
                        "Plugin.json");
        if (embedded.exists()) {
            return Collections.singletonList(new PluginFile(embedded));
        }
        return Collections.emptyList();
    }

    private void load(Plugin plugin) throws IOException {
        plugins.add(plugin);
        if (plugin instanceof Dimension) {
            dimensions.add((Dimension) plugin);
        }
        if (plugin instanceof WorldType) {
            if (worldType != null) {
                throw new IOException("Found 2nd world type: " + plugin);
            }
            worldType = (WorldType) plugin;
        }
    }

    public void dispose() {
        plugins.clear();
        dimensions.clear();
        worldType = null;
        if (classLoader != null) {
            try {
                classLoader.close();
            } catch (IOException e) {
                LOGGER.error("Failed to close plugin classloader: {}",
                        e.toString());
            }
        }
    }

    public GameRegistry registry() {
        return registry;
    }

    public int fileCount() {
        return files.size();
    }

    public Stream<PluginFile> files() {
        return Streams.of(files);
    }

    public PluginFile file(int i) {
        return files.get(i);
    }

    public Stream<Plugin> plugins() {
        return Streams.of(plugins);
    }

    public Stream<Dimension> dimensions() {
        return Streams.of(dimensions);
    }

    public WorldType worldType() {
        return worldType;
    }

    public Plugin plugin(String name) {
        for (Plugin plugin : plugins) {
            if (plugin.id().equals(name)) {
                return plugin;
            }
        }
        throw new IllegalArgumentException("Unknown plugin");
    }

    public void addFileSystems(FileSystemContainer files) {
        for (Plugin plugin : plugins) {
            files.registerFileSystem(plugin.id(),
                    new ClasspathPath(plugin.getClass().getClassLoader(),
                            plugin.assetRoot()));
        }
    }

    public void removeFileSystems(FileSystemContainer files) {
        for (Plugin plugin : plugins) {
            files.removeFileSystem(plugin.id());
        }
    }

    public void init() {
        if (!init) {
            Streams.forEach(plugins, plugin -> plugin.init(registry));
            Streams.forEach(plugins, plugin -> plugin.initEnd(registry));
            registry.lock();
            init = true;
        }
    }
}

File: ScapesCore/src/main/kotlin/org/tobi29/scapes/plugins/Plugins.kt
/*
 * Copyright 2012-2017 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.plugins

import org.tobi29.io.FileSystemContainer
import org.tobi29.io.IOException
import org.tobi29.io.tag.MutableTagMap
import org.tobi29.logging.KLogging
import org.tobi29.scapes.block.*
import org.tobi29.scapes.inventory.ItemType
import org.tobi29.scapes.packets.*
import org.tobi29.scapes.plugins.spi.PluginHandle
import org.tobi29.scapes.plugins.spi.PluginProvider
import org.tobi29.stdex.readOnly
import org.tobi29.utils.spiLoad
import kotlin.reflect.KClass

class Plugins(
    files: List<PluginHandle>,
    idStorage: MutableTagMap
) {
    val files = files.readOnly()
    @PublishedApi
    internal val _plugins: Map<KClass<out Plugin>, Plugin>
    val plugins: Collection<Plugin> get() = _plugins.values
    val dimensions: List<Dimension>
    val worldType: WorldType
    val registry: Registries
    lateinit var air: BlockAir
        private set
    lateinit var materialResolver: Map<String, ItemType>
        private set
    private var init = false

    init {
        val plugins = HashMap<KClass<out Plugin>, Plugin>()
        val dimensions = ArrayList<Dimension>()
        var worldType: WorldType? = null
        for (handle in files) {
            val plugin = handle.second()
            plugins[plugin::class] = plugin
            if (plugin is Dimension) {
                dimensions.add(plugin)
            }
            if (plugin is WorldType) {
                if (worldType != null) {
                    throw IOException("Found 2nd world type: $plugin")
                }
                worldType = plugin
            }
        }
        if (worldType == null) throw IOException("No world type found")
        this._plugins = plugins.readOnly()
        this.dimensions = dimensions.readOnly()
        this.worldType = worldType
        registry = Registries(idStorage)
    }

    fun init() {
        if (!init) {
            registry.registryTypes({ registry ->
                registry.add("Core", "ItemType", 0, Short.MAX_VALUE.toInt())
                registry.add("Core", "Entity", 0, Int.MAX_VALUE)
                registry.add("Core", "Environment", 0, Int.MAX_VALUE)
                registry.add("Core", "Packet", 0, Short.MAX_VALUE.toInt())
                registry.add("Core", "Update", 0, Short.MAX_VALUE.toInt())
                plugins.forEach { it.registryType(registry) }
            })
            init(registry)
            registry.get<ItemType>("Core", "ItemType").run {
                air = reg("core.block.Air", 0) {
                    BlockAir(MaterialType(this@Plugins, it, "core.block.Air"))
                }
            }
            plugins.forEach { it.register(this) }
            plugins.forEach { it.init(registry) }
            registry.lock()
            materialResolver = assembleMaterialResolver(
                registry.get<ItemType>("Core", "ItemType")
            )
            init = true
        }
    }

    inline fun <reified P : Plugin> plugin(): P =
        _plugins[P::class] as? P
                ?: throw IllegalArgumentException("Unknown plugin: ${P::class}")

    companion object : KLogging() {
        private val spiPlugins =
            spiLoad(spiLoad<PluginProvider>()) { e ->
                logger.warn { "Unable to load plugin provider: $e" }
            }.flatMap { it.plugins }.toList()

        fun available(): List<PluginHandle> = spiPlugins

        fun setupAssets(files: FileSystemContainer) {
            available().forEach {
                files.registerFileSystem(it.first.id, it.first.assetRoot)
            }
        }

        fun init(registry: Registries) {
            registry.get<UpdateType>("Core", "Update").run {
                reg("core.update.BlockUpdate") {
                    UpdateType(it, ::UpdateBlockUpdate)
                }
                reg("core.update.BlockUpdateUpdateTile") {
                    UpdateType(it, ::UpdateBlockUpdateUpdateTile)
                }
            }

            registry.get<PacketType>("Core", "Packet").run {
                reg("core.packet.RequestChunk") {
                    PacketType(it, ::PacketRequestChunk)
                }
                reg("core.packet.RequestEntity") {
                    PacketType(it, ::PacketRequestEntity)
                }
                reg("core.packet.SendChunk") {
                    PacketType(it, ::PacketSendChunk)
                }
                reg("core.packet.BlockChange") {
                    PacketType(it, ::PacketBlockChange)
                }
                reg("core.packet.BlockChangeAir") {
                    PacketType(it, ::PacketBlockChangeAir)
                }
                reg("core.packet.EntityAdd") {
                    PacketType(it, ::PacketEntityAdd)
                }
                reg("core.packet.EntityChange") {
                    PacketType(it, ::PacketEntityChange)
                }
                reg("core.packet.EntityComponentData") {
                    PacketType(it, ::PacketEntityComponentData)
                }
                reg("core.packet.EntityMetaData") {
                    PacketType(it, ::PacketEntityMetaData)
                }
                reg("core.packet.MobMoveRelative") {
                    PacketType(it, ::PacketMobMoveRelative)
                }
                reg("core.packet.MobMoveAbsolute") {
                    PacketType(it, ::PacketMobMoveAbsolute)
                }
                reg("core.packet.MobChangeRot") {
                    PacketType(it, ::PacketMobChangeRot)
                }
                reg("core.packet.MobChangeSpeed") {
                    PacketType(it, ::PacketMobChangeSpeed)
                }
                reg("core.packet.MobChangeState") {
                    PacketType(it, ::PacketMobChangeState)
                }
                reg("core.packet.MobDamage") {
                    PacketType(it, ::PacketMobDamage)
                }
                reg("core.packet.EntityDespawn") {
                    PacketType(it, ::PacketEntityDespawn)
                }
                reg("core.packet.SoundEffect") {
                    PacketType(it, ::PacketSoundEffect)
                }
                reg("core.packet.Interaction") {
                    PacketType(it, ::PacketInteraction)
                }
                reg("core.packet.InventoryInteraction") {
                    PacketType(it, ::PacketInventoryInteraction)
                }
                reg("core.packet.PlayerJump") {
                    PacketType(it, ::PacketPlayerJump)
                }
                reg("core.packet.OpenGui") {
                    PacketType(it, ::PacketOpenGui)
                }
                reg("core.packet.CloseGui") {
                    PacketType(it, ::PacketCloseGui)
                }
                reg("core.packet.UpdateInventory") {
                    PacketType(it, ::PacketUpdateInventory)
                }
                reg("core.packet.Chat") {
                    PacketType(it, ::PacketChat)
                }
                reg("core.packet.ItemUse") {
                    PacketType(it, ::PacketItemUse)
                }
                reg("core.packet.Disconnect") {
                    PacketType(it, ::PacketDisconnect)
                }
                reg("core.packet.DisconnectSelf") {
                    PacketType(it, {
                        throw IOException(
                            "This packet should never be received"
                        )
                    })
                }
                reg("core.packet.SetWorld") {
                    PacketType(it, ::PacketSetWorld)
                }
                reg("core.packet.PingClient") {
                    PacketType(it, ::PacketPingClient)
                }
                reg("core.packet.PingServer") {
                    PacketType(it, ::PacketPingServer)
                }
                reg("core.packet.Skin") {
                    PacketType(it, ::PacketSkin)
                }
            }
        }

        fun assembleMaterialResolver(registry: Registries.Registry<ItemType>) =
            registry.values.asSequence().filterNotNull().flatMap { material ->
                val nameID = material.nameID
                sequenceOf(
                    Pair(nameID, material),
                    Pair(
                        nameID.substring(
                            nameID.lastIndexOf('.') + 1,
                            nameID.length
                        ), material
                    )
                )
            }.toMap()
    }
}

fun <T : ItemType> Registries.Registry<ItemType>.reg(
    name: String,
    plugins: Plugins,
    block: (MaterialType) -> T
) = reg(name) { block(MaterialType(plugins, it, name)) }
